internal/resolver/dns/dns_resolver_test.go

```diff
@@ -736,12 +749,122 @@ func testDNSResolver(t *testing.T) {
 	}
 }
 
+// DNS Resolver immediately starts polling on an error from grpc. This should continue until the ClientConn doesn't
+// send back an error from updating the DNS Resolver's state.
+func TestDNSResolverExponentialBackoff(t *testing.T) {
+	defer leakcheck.Check(t)
+	nt := newTimer
+	defer func(func(d time.Duration) *time.Timer) {
+		newTimer = nt
+	}(nt)
+	timerChan := make(chan *time.Timer, 1)
+	newTimer = func(d time.Duration) *time.Timer {
+		// Will never fire on it's own, allows this test to call timer immediately.
+		t := time.NewTimer(time.Hour)
+		timerChan <- t
+		return t
+	}
+	tests := []struct {
+		target   string
+		addrWant []resolver.Address
+		scWant   string
+	}{
+		{
+			"foo.bar.com",
+			[]resolver.Address{{Addr: "1.2.3.4" + colonDefaultPort}, {Addr: "5.6.7.8" + colonDefaultPort}},
+			generateSC("foo.bar.com"),
+		},
+		{
+			"foo.bar.com:1234",
+			[]resolver.Address{{Addr: "1.2.3.4:1234"}, {Addr: "5.6.7.8:1234"}},
+			generateSC("foo.bar.com"),
+		},
+		{
+			"srv.ipv4.single.fake",
+			[]resolver.Address{{Addr: "2.4.6.8" + colonDefaultPort}},
+			generateSC("srv.ipv4.single.fake"),
+		},
+	}
+	for _, a := range tests {
+		b := NewBuilder()
+		cc := &testClientConn{target: a.target}
+		// Cause ClientConn to return an error.
+		cc.updateStateErr = balancer.ErrBadResolverState
+		r, err := b.Build(resolver.Target{Endpoint: a.target}, cc, resolver.BuildOptions{})
+		timer := <-timerChan
+		timer.Reset(0)
+		if err != nil {
+			t.Fatalf("%v\n", err)
+		}
+		var state resolver.State
+		var cnt int
+		for i := 0; i < 2000; i++ {
+			state, cnt = cc.getState()
+			if cnt > 0 {
+				break
+			}
+			time.Sleep(time.Millisecond)
+		}
+		if cnt == 0 {
+			t.Fatalf("UpdateState not called after 2s; aborting")
+		}
+		if !reflect.DeepEqual(a.addrWant, state.Addresses) {
+			t.Errorf("Resolved addresses of target: %q = %+v, want %+v", a.target, state.Addresses, a.addrWant)
+		}
+		sc := scFromState(state)
+		if a.scWant != sc {
+			t.Errorf("Resolved service config of target: %q = %+v, want %+v", a.target, sc, a.scWant)
+		}
+		// Cause timer to go off 10 times, and see if it calls updateState() correctly.
+		for i := 0; i < 10; i++ {
+			timer := <-timerChan
+			timer.Reset(0)
+			// Propagate.
+			time.Sleep(time.Millisecond)
```

Comment:
What does this propagate it *to*?  And why is 1ms considered long enough, always?

In general, please try to avoid any sleeps at all.

Reply:
Changed to polling logic.